File: Data Structure Learning/Concurency/Leetcode problem/Printorder.java
import java.util.concurrent.Semaphore;

public class Printorder {
    private Semaphore semaSec = new Semaphore(0);
    private Semaphore semaThird = new Semaphore(0);

    public Printorder(){}

    public void first(Runnable printFirst) throws InterruptedException{
        printFirst.run(); 
        semaSec.release();
    }

    public void second(Runnable printSecond) throws InterruptedException{
        semaSec.acquire();
        printSecond.run();
        semaSec.release();
    }

    public void third(Runnable printThird) throws InterruptedException{
        semaThird.acquire();
        printThird.run();
    
    }
    
}

File: Data Structure Learning/Concurency/Leetcode problem/Printorder.py
from threading import Semaphore

class Printorder:
    def __init__(self):
        self.second_ready = Semaphore(0)
        self.third_ready = Semaphore(0)
    
    def first(self, printFirst):
        printFirst()
        self.second_ready.release()
    
    def second(self, printSecond):
        self.second_ready.acquire()
        printSecond()
        self.third_ready.release()

    def third(self, printThird):
        self.third_ready.acquire()
        printThird()
        self.third_ready.release()